lib/iris/experimental/ugrid/__init__.py

```diff
@@ -2746,6 +2747,239 @@ def face_node(self):
+        constructed from the :attr:`~iris.coords.Coord.bounds` of ``coord_1``
+        and ``coord_2``. :attr:`~Mesh.face_coords` constructed from the
+        :attr:`~iris.coords.Coord.points` of ``coord_1`` and ``coord_2``.
+
+    Kwargs:
+
+    * coord_1 (:class:`~iris.coords.Coord`):
+        The coordinate corresponding to the :class:`Mesh`'s first axis.
+
+    * coord_2 (:class:`~iris.coords.Coord`):
+        The coordinate corresponding to the  :class:`Mesh`'s second axis. Must
+        be orthogonal to ``coord_1``. :attr:`~iris.coords.Coord.points` and
+        :attr:`~iris.coords.Coord.bounds` must have same shapes as those of
+        ``coord_1``.
+
+    Returns:
+        :class:`Mesh`
+
+    .. note::
+        Any resulting duplicate nodes are not currently removed, due to the
+        computational intensity.
+
+    .. testsetup::
+
+        from iris import load_cube, sample_data_path
+        from iris.experimental.ugrid import (
+            PARSE_UGRID_ON_LOAD,
+            MeshCoord,
+            mesh_from_coords,
+        )
+
+        file_path = sample_data_path("mesh_C4_synthetic_float.nc")
+        with PARSE_UGRID_ON_LOAD.context():
+            cube_w_mesh = load_cube(file_path)
+
+    For example::
+
+        # Reconstruct a cube-with-mesh after subsetting it.
+
+        >>> print(cube_w_mesh.mesh.name())
+        Topology data of 2D unstructured mesh
+        >>> mesh_coord_names = [
+        ...     coord.name() for coord in cube_w_mesh.coords(mesh_coords=True)
+        ... ]
+        >>> print(f"MeshCoords: {mesh_coord_names}")
+        MeshCoords: ['latitude', 'longitude']
+
+        # Subsetting converts MeshCoords to AuxCoords.
+        >>> slices = [slice(None)] * cube_w_mesh.ndim
+        >>> slices[cube_w_mesh.mesh_dim()] = slice(-1)
+        >>> cube_sub = cube_w_mesh[tuple(slices)]
+        >>> print(cube_sub.mesh)
+        None
+        >>> orig_mesh_coords = [cube_sub.coord(c_name) for c_name in mesh_coord_names]
+        >>> for coord in orig_mesh_coords:
+        ...     print(f"{coord.name()}: {type(coord).__name__}")
+        latitude: AuxCoord
+        longitude: AuxCoord
+
+        >>> new_mesh = mesh_from_coords(orig_mesh_coords[0], orig_mesh_coords[1])
+        >>> new_mesh_coords = new_mesh.to_MeshCoords(location=cube_w_mesh.location)
+
+        # Replace the AuxCoords with MeshCoords.
+        >>> for ix in range(2):
+        ...     cube_sub.remove_coord(orig_mesh_coords[ix])
+        ...     cube_sub.add_aux_coord(new_mesh_coords[ix], cube_w_mesh.mesh_dim())
+
+        >>> print(cube_sub.mesh.name())
+        Topology data of 2D unstructured mesh
+        >>> for coord_name in mesh_coord_names:
+        ...     coord = cube_sub.coord(coord_name)
+        ...     print(f"{coord_name}: {type(coord).__name__}")
+        latitude: MeshCoord
+        longitude: MeshCoord
+
+    """
+    # Validate coord essentials.
+    coord_names = {coord_1: "coord_1", coord_2: "coord_2"}
```

Comment:
It seems like bad practice to use a mutable object as a key in a dictionary. It might make more sense to reverse the order of key/item here

Reply:
Well, Coord is one of those classes [specifically adjusted](https://github.com/SciTools/iris/blob/v3.1.0/lib/iris/coords.py#L1610-L1617) so that we _can_ use them as dictionary keys and sets.

Frankly, the more I see of this, the less it worries me -- it's already a kluge in Python itself that generic objects can be used in this way.
And I think it basically just makes `dict[obj] = value` a shorthand for `dict[id(obj)] = value`.

Reply:
>It seems like bad practice to use a mutable object as a key in a dictionary.

Pretty sure `Coord`s are immutable. Indeed that's why I was comfortable with this, but it sounds like that's irrelevant anyway!